pkg/testcase/clusterrestore.go

```diff
@@ -0,0 +1,403 @@
+	s3, err := aws.AddClient(cluster)
+	Expect(err).NotTo(HaveOccurred(), "error creating s3 client: %s", err)
+
+	s3List, s3ListErr := s3.GetObjects(flags)
+	Expect(s3ListErr).NotTo(HaveOccurred())
+	for _, listObject := range s3List {
+		if strings.Contains(*listObject.Key, onDemandPath) {
+			shared.LogLevel("info", "snapshot found: %s", onDemandPath)
+			break
+		}
+	}
+}
+
+func stopInstances(cluster *shared.Cluster, ec2 *aws.Client) {
+	var instancesIPs []string
+
+	instancesIPs = append(instancesIPs, cluster.ServerIPs...)
+	instancesIPs = append(instancesIPs, cluster.AgentIPs...)
+
+	for _, ip := range instancesIPs {
+		id, idsErr := ec2.GetInstanceIDByIP(ip)
+		Expect(idsErr).NotTo(HaveOccurred())
+		//
+		err := ec2.StopInstance(id)
+		if err != nil {
+			return
+		}
+	}
+}
+
+func setConfigFile(product, newClusterIP string) {
+	var createConfigFileCmd string
+
+	if product == "rke2" {
+		createConfigFileCmd = fmt.Sprintf("sudo  bash -c 'cat <<EOF>/etc/rancher/%s/config.yaml\n"+
+			"write-kubeconfig-mode: 644\n"+
+			"node-external-ip: %s\n"+
+			os.Getenv("server_flags")+"\n"+
+			"EOF'", product, newClusterIP)
+	} else {
+		createConfigFileCmd = fmt.Sprintf("sudo  bash -c 'cat <<EOF>/etc/rancher/%s/config.yaml\n"+
+			"write-kubeconfig-mode: 644\n"+
+			"node-external-ip: %s\n"+
+			"cluster-init: true\n"+
+			"EOF'", product, newClusterIP)
+	}
+
+	path := fmt.Sprintf("/etc/rancher/%s/", product)
+	cmd := fmt.Sprintf("sudo  mkdir -p %s && %s", path, createConfigFileCmd)
+
+	_, mkdirCmdErr := shared.RunCommandOnNode(cmd, newClusterIP)
+	Expect(mkdirCmdErr).NotTo(HaveOccurred())
+}
+
+func installProduct(
+	cluster *shared.Cluster,
+	newClusterIP string,
+	version string,
+) {
+	setConfigFile(cluster.Config.Product, newClusterIP)
+
+	installCmd := "curl -sfL "
+	if cluster.Config.Product == "k3s" {
+		installCmd += fmt.Sprintf("https://get.%s.io/ | sudo INSTALL_%s_VERSION=%s  INSTALL_%s_SKIP_ENABLE=true sh -",
+			cluster.Config.Product, strings.ToUpper(cluster.Config.Product), version,
+			strings.ToUpper(cluster.Config.Product))
+	} else {
+		installCmd += fmt.Sprintf("https://get.%s.io | sudo INSTALL_%s_VERSION=%s sh -",
+			cluster.Config.Product, strings.ToUpper(cluster.Config.Product), version)
+	}
+
+	_, installCmdErr := shared.RunCommandOnNode(installCmd, newClusterIP)
+	Expect(installCmdErr).NotTo(HaveOccurred())
+}
+
+func enableAndStartService(
+	cluster *shared.Cluster,
+	newClusterIP string,
+) {
+	_, enableServiceCmdErr := shared.ManageService(cluster.Config.Product, "enable", "server",
+		[]string{newClusterIP})
+	Expect(enableServiceCmdErr).NotTo(HaveOccurred())
+	_, startServiceCmdErr := shared.ManageService(cluster.Config.Product, "start", "server",
+		[]string{newClusterIP})
+
+	shared.LogLevel("info", "Starting service, waiting for service to complete background processes.")
+	Expect(startServiceCmdErr).NotTo(HaveOccurred())
+
+	time.Sleep(300 * time.Second)
```

Comment:
we dont need that much

also lets add a TODO: comment on all sleeps like

//TODO: improve/remove sleeps with k8s client  ( something like that. )